File: backend/src/main/java/com/main/Difficult/Domain/producto/compuesto/Combo.java
package com.main.Difficult.Domain.producto.compuesto;

import com.fasterxml.jackson.annotation.JsonView;
import com.main.Difficult.Domain.producto.Lote;
import com.main.Difficult.Domain.producto.Producto;
import com.main.Difficult.Domain.producto.View;
import com.main.Difficult.Domain.producto.simple.ProductoSimple;
import org.springframework.data.annotation.TypeAlias;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.atomic.AtomicReference;

@Document("Producto")
@TypeAlias("Combo")
public class Combo extends Producto {
    @JsonView(View.ProductoItem.class)
    private Set<ItemCombo> items = new HashSet<>();

    public Combo(String imagen, String nombre, String descripcion, Integer puntaje, String paisDeOrigen, Double precioBase) {
        super(imagen, nombre, descripcion, puntaje, paisDeOrigen, precioBase);
    }

    public void agregarProductoCombo(ProductoSimple item , Lote lote, Integer cantidad){
        items.add(new ItemCombo(item,lote,cantidad));
    }
    public void borrarProductoCombo(ProductoSimple item){
        items.removeIf((_item)-> _item.getProducto().getId() == item.getId());
    }
    @Override
    public Boolean hayProductoConMasDe4Meses(){
        return super.hayProductoConMasDe4Meses() || this.hayItemsConMasDe4Meses();
    }
    public Boolean hayItemsConMasDe4Meses(){
        return items.stream().anyMatch((item)-> item.tieneMasDe4Meses());
    }
    public Double precio(){
        AtomicReference<Double> itemsCalculoPrecio = new AtomicReference<>(0.0);
        items.forEach((item) -> itemsCalculoPrecio.updateAndGet(v -> v + item.precioBase()));
        return (0.85 * (itemsCalculoPrecio.get() + 20 * items.size())) *this.descuento();

    }


}

File: backend/src/main/kotlin/aplicacion/dominio/producto/compuesto/Combo.kt
package aplicacion.dominio.producto.compuesto

import aplicacion.dominio.Item
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonView
import aplicacion.dominio.producto.Lote
import aplicacion.dominio.producto.Producto
import aplicacion.dominio.producto.View
import aplicacion.dominio.producto.simple.ProductoSimple
import javax.persistence.*

@Entity
class Combo : Producto() {
    @JsonView(View.ProductoIndividual::class)
    @OneToMany(cascade = [CascadeType.MERGE,CascadeType.PERSIST], fetch = FetchType.EAGER) @JoinColumn(name = "combo_id")
    @JsonIgnore var items: MutableSet<ItemCombo> = mutableSetOf()

    fun agregarProductoCombo(item: ProductoSimple, lote: Lote, cantidad: Int) {
        items.add(ItemCombo(item, lote, cantidad))
    }

    fun borrarProductoCombo(item: ProductoSimple) {
        items.removeIf { it.producto.id == item.id }
    }

    override fun hayProductoConMasDe4Meses(): Boolean {
        return super.hayProductoConMasDe4Meses() || hayItemsConMasDe4Meses()
    }

    private fun hayItemsConMasDe4Meses(): Boolean {
        return items.any { it.tieneMasDe4Meses() }
    }

    override fun precio(): Double {
        return (0.85 * (items.sumOf { it.precioBase() } + 20 * items.size)) * descuento()
    }
}

/* PRODUCTOSIMPLE - LOTE - COMBO */
@JsonView(View.ProductoIndividual::class)
@Entity @Table(name="combo_producto_lote")
class ItemCombo():Item(){

    constructor(producto: ProductoSimple, lote: Lote, cantidad: Int) : this() {
        this.producto = producto
        this.lote = lote
        this.cantidad = cantidad
    }

    fun precioBase(): Double {
        return producto.precioBase
    }

    fun tieneMasDe4Meses(): Boolean {
        return lote.tieneMasDe4Meses()
    }

    @JsonProperty fun producto(): String = producto.nombre
    @JsonProperty("Precio") fun productoPrecio(): String = "$"+producto.precioBase.toString()
    @JsonProperty fun lote(): String = lote.id.toString()
}